File: BinaryTreePostorderTraversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<Integer>();
        Stack<TreeNode> s = new Stack<TreeNode>();
        TreeNode cur = root, p = null;
        
        while (cur != null || !s.isEmpty()) {
            while (cur != null) {
                s.push(cur);
                cur = cur.left;
            }
            cur = s.pop();
            if (cur.right == null || cur.right == p) {
                list.add(cur.val);
                p = cur;
                cur = null;
            } else {
                s.push(cur);
                cur = cur.right;
            }
        }
        return list;
    }
}

File: lulu/binaryTreePostorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.ans = []
        
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        self.postorderHelper(root)
        return self.ans
    
    def postorderHelper(self, root):
        if root:      
            self.postorderHelper(root.left)
            self.postorderHelper(root.right)
            self.ans.append(root.val)
